### DISCUSSION TITLE
Negation in applicability

### DISCUSSION THREAD

**Discussion by atomczak** [buildingSMART/IDS#347] [Q&A]
I want to have a requirement that takes all Walls that don't have the "FireRating" property. In other words, I want an applicability to include IFCWALL entity and a property allowing any name/pset except "FireRating" and "Pset_WallCommon". I thought I could do that with a pattern, but I see that XML flavour doesn't really support a negation lookahead. Has anyone found a way to express it?

**Comment by NickNisbet** [comment]
Add a clause called Exception with all the usual facets.

**Reply by atomczak** [reply]
That could work, but that means changing the schema and increasing its complexity. And it stretches the scope of IDS to design rules. I'm asking if there is a solution to this with IDS1.0. If we are to make changes, I would rather suggest reconsidering regex flavour, as XSD is simply not made for our purposes. I would replace it with more common flavour like PCRE, JavaScript or Python, that allow negations.

**Comment by andyward** [comment]
Have hit this as well. I think we used to be able to prohibit facets in applicability with `maxOccurs=0`. But this got lost in the IDS0.96/97 refactoring when 'cardinality' was introduced - but only on requirement facets. I don't see why we can't have a form of cardinality on the applicability facets.

I've used negation regexs as a workaround. It's not clear to me artificially restricted the regex flavour. As this is just an implementors agreement it would seem relatively simple to change in 1.0 vs a schema update.

E.g. find all entities except those with names starting 'IGNORE_'

```xml
...
  </ids:entity>
  <ids:attribute>
	  <ids:name>
		  <ids:simpleValue>Name</ids:simpleValue>
	  </ids:name>
	  <ids:value>
		  <xs:restriction base="xs:string">
			  <xs:pattern value="^(?!IGNORE_.*$).*" />
		  </xs:restriction>
	  </ids:value>
  </ids:attribute>
</ids:applicability>
```

**Reply by atomczak** [reply]
As of now, that won't work - ^, $, ?! are literal characters, not not start/end of the string etc. We would need to agree on another regex flavour altogether.

Let's raise it when the group restarts (the plan is mid-October).
